action/governance/withdrawFunds.go

```diff
@@ -0,0 +1,234 @@
+	return runWithdraw(ctx, tx)
+}
+
+func (wp WithdrawFunds) ProcessDeliver(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Debug("Processing WithdrawFunds Transaction for DeliverTx", tx)
+	return runWithdraw(ctx, tx)
+}
+
+func (wp WithdrawFunds) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 1)
+}
+
+func runWithdraw(ctx *action.Context, signedTx action.RawTx) (bool, action.Response) {
+	withdrawProposal := WithdrawFunds{}
+	err := withdrawProposal.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, action.Response{}
+	}
+
+	// 1. Check if Proposal already exists, if so, check the withdraw requirement:
+	//    a. the funding goal is not reached
+	//    b. the funding height is reached
+
+	proposal, _, err := ctx.ProposalMasterStore.Proposal.QueryAllStores(withdrawProposal.ProposalID)
+	if err != nil {
+		ctx.Logger.Error("Proposal does not exist :", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_does_not_exist"),
+		}
+		return false, result
+	}
+	currentFundsForProposal := governance.GetCurrentFunds(proposal.ProposalID, ctx.ProposalMasterStore.ProposalFund)
+	// if funding goal is reached or there is still time for funding
+	if currentFundsForProposal.BigInt().Cmp(proposal.FundingGoal.BigInt()) >= 0 || ctx.Header.Height <= proposal.FundingDeadline {
+		ctx.Logger.Error("Proposal does not meet withdraw requirement", withdrawProposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "withdraw_proposal_does_not_meet_withdraw_requirement"),
+		}
+		return false, result
+	}
+	// 2. change outcome, status, state
+	proposal.Outcome = governance.ProposalOutcomeInsufficientFunds
+	proposal.Status = governance.ProposalStatusCompleted
+	err = ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateFailed).Set(proposal)
+	if err != nil {
+		ctx.Logger.Error("Failed to add proposal to FAILED store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_add_proposal_to_failed_store"),
+		}
+		return false, result
+	}
+	ok, err := ctx.ProposalMasterStore.Proposal.WithPrefixType(governance.ProposalStateActive).Delete(proposal.ProposalID)
+	if err != nil || !ok {
+		ctx.Logger.Error("Failed to delete proposal from ACTIVE store :", proposal.ProposalID)
+		result := action.Response{
+			Events: action.GetEvent(withdrawProposal.Tags(), "failed_to_delete_proposal_from_active_store"),
+		}
+		return false, result
+	}
+
+	// 3. Check if the contributor has funded this proposal, if so, get the amount of funds
+	proposalFund, err := governance.GetCurrentFundsByContributor(proposal.ProposalID, withdrawProposal.Contributor, ctx.ProposalMasterStore.ProposalFund)
```

Comment:
If all we need to do here is to 'WithdrawFund' and this is the only place func 'GetCurrentFundsByContributor' is used. We actually can move this func's logic into DeductFunds(). This will make the logic here very simple. But it works here and it's up to you :).

Reply:
yeah, I think put it here will return the error will clearly points out that the contributor has not funded the proposal. and the error message from deduct will only say insufficient funds.